### PR TITLE
Avoid WebRequest deadlock in Telemetry

### PR DESCRIPTION
**Pull request by tonyredondo**
In some scenarios the Telemetry worker triggers a deadlock when using the HttpWebRequest implementation (.NET Framework) on process shutdown.

The deadlock appear to be a bug in the HttpWebRequest implementation according to: https://social.msdn.microsoft.com/Forums/en-US/2ab06a91-c8fd-474c-81ea-61b78c8fbb56/httpwebrequestgetrequeststream-hangs-bug-in-systemnet-stack-?forum=ncl

To avoid the deadlock, the `JsonWebRequestTelemetryTransport` implementation changed from Async to Sync  and wrapped by a `lock` mechanism, as suggested by the forum response:

image

### CODE REVIEW

File: tracer/src/Datadog.Trace/LifetimeManager.cs [hunk 1]
```diff
@@ -75,12 +76,14 @@ private void CurrentDomain_UnhandledException(object sender, UnhandledExceptionE
 
         private void Console_CancelKeyPress(object sender, ConsoleCancelEventArgs e)
         {
+            Log.Debug("Running Shutdown tasks due CancelKeyPress");
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/LifetimeManager.cs:79]:
```suggestion
            Log.Debug("Running Shutdown tasks due to CancelKeyPress");
```

File: tracer/src/Datadog.Trace/LifetimeManager.cs [hunk 2]
```diff
@@ -75,12 +76,14 @@ private void CurrentDomain_UnhandledException(object sender, UnhandledExceptionE
...
             RunShutdownTasks();
             Console.CancelKeyPress -= Console_CancelKeyPress;
         }
 
         private void CurrentDomain_DomainUnload(object sender, EventArgs e)
         {
+            Log.Debug("Running Shutdown tasks due DomainUnload");
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/LifetimeManager.cs:86]:
```suggestion
            Log.Debug("Running Shutdown tasks due to DomainUnload");
```

File: tracer/src/Datadog.Trace/TracerManager.cs [hunk 3]
```diff
@@ -439,17 +439,49 @@ private static void OneTimeSetup()
...
+                    _heartbeatTimer?.Dispose();
+
+                    var instance = _instance;
+                    if (instance is not null)
+                    {
+                        Log.Debug("Shutdown agent writer.");
+                        await _instance.AgentWriter.FlushAndCloseAsync().ConfigureAwait(false);
+
+                        Log.Debug("Shutdown direct log submission.");
+                        _instance.DirectLogSubmission?.Dispose();
```
Comment (andrewlock) [inline tracer/src/Datadog.Trace/TracerManager.cs:471]: I wonder if there's a similar path in here we need to make async?

**Comment by tonyredondo** [comment]
The deadlock issue we have seen was not related to the forum post, but for missing the disposal of every telemetry response.

So, this PR has been superseeded by https://github.com/DataDog/dd-trace-dotnet/pull/2574

**Pull request closed without merging**